specs/phase0/beacon-chain.md

```diff
@@ -601,17 +605,51 @@ def bytes_to_uint64(data: bytes) -> uint64:
+##### IETF standard
 
-- `def Sign(SK: int, message: Bytes) -> BLSSignature`
-- `def Verify(PK: BLSPubkey, message: Bytes, signature: BLSSignature) -> bool`
+This specification refers to [draft v4 of the IETF BLS signature standard](https://tools.ietf.org/html/draft-irtf-cfrg-bls-signature-04). Specifically, the following endpoints are used with the `BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_POP_` ciphersuite:
+
+- `def Sign(secret_key: int, message: bytes) -> BLSSignature`
 - `def Aggregate(signatures: Sequence[BLSSignature]) -> BLSSignature`
-- `def FastAggregateVerify(PKs: Sequence[BLSPubkey], message: Bytes, signature: BLSSignature) -> bool`
-- `def AggregateVerify(PKs: Sequence[BLSPubkey], messages: Sequence[Bytes], signature: BLSSignature) -> bool`
+- `def _Verify(pubkey: BLSPubkey, message: bytes, signature: BLSSignature) -> bool`
+- `def _AggregateVerify(pubkeys: Sequence[BLSPubkey], messages: Sequence[bytes], signature: BLSSignature) -> bool`
+- `def _FastAggregateVerify(pubkeys: Sequence[BLSPubkey], message: bytes, signature: BLSSignature) -> bool`
+
+For notational clarity the above endpoints are accessed through the `ietf` module, e.g. `eitf_v4._AggregateVerify`.
+
+##### Eth2 wrappers
+
+Eth2 wraps the above endpoints from draft v4 of the IETF BLS signature standard to make two changes:
+
+- allow infinity pubkeys (replicating [draft v3](https://tools.ietf.org/html/draft-irtf-cfrg-bls-signature-03) for historical reasons)
+- allow empty aggregate signatures for phases 1+
+
+```python
+def bls_aggregate_verify(pubkeys: Sequence[BLSPubkey], messages: Sequence[bytes], signature: BLSSignature) -> bool:
+    if len(pubkeys) == 0:
+        return False
 
-Within these specifications, BLS signatures are treated as a module for notational clarity, thus to verify a signature `bls.Verify(...)` is used.
+    # Filter out (pubkey, message) pairs with an infinity pubkey
+    pubkeys_and_messages = [(pubkey, message) for pubkey, message in zip(pubkeys, messages)
+                            if pubkey != G1_INFINITY_POINT]
+    # If all the pubkeys are infinity pubkeys check the signature is the infinity signature
+    if len(pubkeys_and_messages) == 0:
+        return signature == G2_INFINITY_POINT
+    pubkeys, messages = list(zip(*pubkeys_and_messages))
+    return ietf._AggregateVerify(pubkeys, messages, signature)
+```
+
+```python
+def bls_verify(pubkey: BLSPubkey, message: bytes, signature: BLSSignature) -> bool:
+    return bls_aggregate_verify([pubkey], [message], signature)
+```
+
+```python
+def bls_fast_aggregate_verify(pubkeys: Sequence[BLSPubkey], message: bytes, signature: BLSSignature) -> bool:
+    return bls_aggregate_verify(pubkeys, [message] * len(pubkeys), signature)
```

Comment:
@JustinDrake 

I see the benefit of cleaner code in the spec. But isn't it less efficient by adding many unnecessary `hash_to_point`?

Reply:
If clients are supposed to implement a wrapper like this, then I would make a best-effort to keep the optimized version around. Otherwise you end up having a bigger divergence between spec and client, making things worse.